### PR TITLE
feat: launch preinstalled WDA process without xcodebuild for a real device with usePreinstalledWDA

### PR DESCRIPTION
**Pull request by KazuCocoa**
We can start a pre installed WDA package with `Xctest` https://github.com/appium/appium-ios-device#classes

This PR depends on https://github.com/appium/WebDriverAgent/pull/687. Once the WDA PR has merged and bump the version, like below capability will allow us to start preinstalled WDA without xcodebuild for a real device.

The WDA can be non-paid account's package.

```
{
  "appium:automationName": "xcuitest",
  "platformName": "ios",
  "appium:platformVersion": "15.5",
  "appium:usePreinstalledWDA": true,
  "appium:updatedWDABundleId": "com.kazucocoa.WebDriverAgentRunner.xctrunner",
  "appium:udid": "<udid>"
}
```

### CODE REVIEW

**Comment by KazuCocoa** [comment]
mm... merge master broke this branch...

**Comment by KazuCocoa** [comment]
Latest master's ci is failing in npm ci as no package-lock.json. 
local result:
```
WARN XCUITest 'undefined' cannot be parsed as a valid URL
      ✔ should be false with invalid input, undefined
      ✔ should be false with invalid input, empty
      ✔ should be true with ipv4 localhost
      ✔ should be true with ipv4 localhost with port
      ✔ should be true with ipv4 127.0.0.1
      ✔ should be true with ipv6 ::1
      ✔ should be true with ipv6 ::ffff:127.0.0.1
      ✔ should be true with ipv6 ::ffff:127.0.0.1 with port
      ✔ should be false with ipv4 192.168.1.100
      ✔ should be false with ipv4 192.168.1.100 with port
      ✔ should be false with ipv6 2001:db8:85a3:8d3:1319:8a2e:370:7348


  437 passing (4s)
```

File: lib/driver.js [hunk 1]
```diff
@@ -866,6 +886,42 @@ class XCUITestDriver extends BaseDriver {
...
+      if (this.opts.wdaLocalPort) {
+        xctestEnv.USE_PORT = this.opts.wdaLocalPort;
+      }
+      this.xctestWDA = new Xctest(this.opts.udid, launchWDABundleID, null, {env: xctestEnv});
+    }
+    await this.xctestWDA.start();
+
+    this.wda.setupProxies(sessionId);
+    this.wda.fullyStarted = true;
+    return await this.wda.getStatus();
```
Comment (KazuCocoa) [inline lib/driver.js:910]: I'll try to move this part in https://github.com/appium/WebDriverAgent (need to add appium-ios-device dependency tho) to wrap around this in `this.wda`

File: docs/capabilities.md [hunk 2]
```diff
@@ -56,6 +56,8 @@ Capability | Description
 |`appium:showXcodeLog`|Whether to display the output of the Xcode command used to run the tests. If this is `true`, there will be **lots** of extra logging at startup. Defaults to `false`|e.g., `true`|
 |`appium:iosInstallPause`|Time in milliseconds to pause between installing the application and starting `WebDriverAgent` on the device. Used particularly for larger applications. Defaults to `0`|e.g., `8000`|
 |`appium:usePrebuiltWDA`|Skips the build phase of running the WDA app. Building is then the responsibility of the user. Only works for Xcode 8+. Defaults to `false`.|e.g., `true`|
+|`appium:usePreinstalledWDA`| Whether to launch a preinstalled WebDriverAgentRunner bundle id instead of running `xcodebuild` to start a WebDriverAgentRunner process. The default target bundle id is `com.facebook.WebDriverAgentRunner.xctrunner`. `appium:preInstalledWDABundleId` will customize the target. Defaults to `false`. |`true` or `true`|
```
Comment (mykola-mokhnach) [inline docs/capabilities.md:59]: I assume we need a dedicated document that explains how to setup WDA, so it runs without xcodebuild

Comment (mykola-mokhnach) [reply docs/capabilities.md:59]: and link to it here

File: docs/real-device-config.md [hunk 3]
```diff
@@ -222,6 +222,30 @@ This would make the device management more flexible, but you'd need to know abou
 > It is recommended to sign packages with a wildcard (`*`) provisioning profile. Such profiles require a paid Apple Developer account. In case of a free account, you may need to update bundle id properly before building the WebDriverAgent package.
 
 
+### Runs preinstalled WebDriverAgent
```
Comment (mykola-mokhnach) [inline docs/real-device-config.md:225]: It makes sense to have a separate document for this tutorial

File: docs/real-device-config.md [hunk 4]
```diff
@@ -222,6 +222,30 @@ This would make the device management more flexible, but you'd need to know abou
...
+
+`appium:usePreinstalledWDA` capability allows Appium to start a preinstalled WebDriverAgentRunner process without `xcodebuild`.
+The preinstalled WebDriverAgent can be signed with non-paid account.
```
Comment (mykola-mokhnach) [inline docs/real-device-config.md:228]: I would prefer to have a separate step-by-step tutorial on how to preinstall WDA on the device under test

File: lib/desired-caps.js [hunk 5]
```diff
@@ -2,317 +2,324 @@
 const PLATFORM_NAME_IOS = 'iOS';
 const PLATFORM_NAME_TVOS = 'tvOS';
 
-const desiredCapConstraints = /** @type {const} */({
-  platformName: { // override
+const desiredCapConstraints = /** @type {const} */ ({
+  platformName: {
```
Comment (mykola-mokhnach) [inline lib/desired-caps.js:6]: please revert these automatic changes. They prevent seeing what's really changed. We could extract them into a separate PR if needed

File: lib/driver.js [hunk 6]
```diff
@@ -465,6 +466,13 @@ class XCUITestDriver extends BaseDriver {
         }/welcome`;
   }
 
+  /**
+   * @returns {boolean} if 'usePreinstalledWDA' capability is enabled.
+   */
+  usePreinstalledWDA() {
```
Comment (mykola-mokhnach) [inline lib/driver.js:472]: should it be a getter?

File: lib/driver.js [hunk 7]
```diff
@@ -818,7 +837,20 @@ class XCUITestDriver extends BaseDriver {
...
               `and reboot the device.`;
           }
-          await quitAndUninstall(errorMsg);
+          if (this.usePreinstalledWDA()) {
+            // In case the bundle id process start got failed because of
+            // auth popup in the device. Then, the bundle id process itself started. It is safe to stop it here.
+            await this.mobileKillApp(this.wda.bundleIdForXctest());
+            // Mostly it failed to start the WDA process as no the bundle id
+            // e.g. '<bundle id of WDA> not found on device <udid>'
+            errorMsg =
```
Comment (mykola-mokhnach) [inline lib/driver.js:846]: do we need to redefine errorMsg here?

File: lib/desired-caps.js (hunk 5 above)
Comment (KazuCocoa) [reply lib/desired-caps.js:6]: Yea... let me skip the commit hook to change this file

File: lib/driver.js (hunk 7 above)
Comment (KazuCocoa) [reply lib/driver.js:846]: yea, since existing message explained xcodebuild failure something, which could confuse the user as this use case.

Comment (mykola-mokhnach) [reply lib/driver.js:846]: but how? It is not used below anywhere (except of the `else` block)

Comment (KazuCocoa) [reply lib/driver.js:846]: Do you mean giving directly or define a new const for `throw new Error(errorMsg);` in L850 instead of using existing `errorMsg`?

File: lib/driver.js [hunk 8]
```diff
@@ -792,18 +800,26 @@ class XCUITestDriver extends BaseDriver {
...
           this.log.info(`Retrying WDA startup (${retryCount + 1} of ${startupRetries})`);
         }
         try {
-          // on xcode 10 installd will often try to access the app from its staging
+          if (usePreinstalledWDA) {
+            // Stop the existing process before starting a new one to start a fresh WDA process every session.
+            await this.mobileKillApp(this.wda.bundleIdForXctest);
+
+            // TODO: Define `prebuiltWDAPath` or something as a new capability.
+            // Then, we can start installing the WDA package to start a session without xcodebuild.
```
Comment (KazuCocoa) [inline lib/driver.js:815]: I'll do this idea as another PR. Then, maybe we can guide users to prepare a wda package and use it (without preinstalling it as well)

**Comment by KazuCocoa** [comment]
This is ready after the wda merge and update

File: docs/capabilities.md [hunk 9]
```diff
@@ -56,6 +56,8 @@ Capability | Description
 |`appium:showXcodeLog`|Whether to display the output of the Xcode command used to run the tests. If this is `true`, there will be **lots** of extra logging at startup. Defaults to `false`|e.g., `true`|
 |`appium:iosInstallPause`|Time in milliseconds to pause between installing the application and starting `WebDriverAgent` on the device. Used particularly for larger applications. Defaults to `0`|e.g., `8000`|
 |`appium:usePrebuiltWDA`|Skips the build phase of running the WDA app. Building is then the responsibility of the user. Only works for Xcode 8+. Defaults to `false`.|e.g., `true`|
+|`appium:usePreinstalledWDA`| Whether to launch a preinstalled WebDriverAgentRunner bundle id instead of running `xcodebuild` to start a WebDriverAgentRunner process. `appium:webDriverAgentUrl` is prior than this capability. The default target bundle id is `com.facebook.WebDriverAgentRunner.xctrunner`. `appium:preInstalledWDABundleId` will customize the target. Please read [Run Preinstalled WebDriverAgentRunner](./run-preinstalled-wda.md) for more details. Defaults to `false`. |`true` or `true`|
```
Comment (mykola-mokhnach) [inline docs/capabilities.md:59]: should we mention it only works for real devices?

File: docs/run-preinstalled-wda.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
+---
+title: Run Preinstalled WebDriverAgentRunner
+---
+
+XCUITest driver can launch preinstalled WebDriverAgent directly against a real device.
+It lets you start an Appium session without the `xcodebuild` command execution to improve the session start performance.
+
+`appium:webDriverAgentUrl` let you manage everything related to the WebDriverAgent, but this functionality let Appium to manage instead.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:8]: I don't understand what this sentence is trying to say. Could you please explain?

File: docs/run-preinstalled-wda.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+## For Real Device
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:10]: Real Devices

File: docs/run-preinstalled-wda.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+### Capabilities
+
+- Required
+  - `appium:usePreinstalledWDA` to start a WebDriverAgent process without xcodebuild
+- Optional
+  - `appium:preInstalledWDABundleId` to customize the WebDriverAgent bundle id Appium launches
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:17]: Lets not use Appium, but XCUITest driver everywhere. Appium is just a server and how nothing to do with iOS automation itself.

File: docs/run-preinstalled-wda.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  "appium:preInstalledWDABundleId": "com.appium.WebDriverAgentRunner.xctrunner"
+}
+@core = Appium::Core.for capabilities: capabilities
+@core.start_driver
+```
+
+If the `<udid>` device has a WebDriverAgent package with `com.appium.WebDriverAgentRunner.xctrunner` bundle id, the session will launch the WebDriverAgent process without xcodebuild.
+
+> **Note**
+> Please ensure that the bundle id is launchable before starting an Appium session.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:41]: ensure that the `WebDriverAgent-Runner` application is launchable ...

File: docs/run-preinstalled-wda.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+Please read [Real Device Configuration tutorial](real-device-config.md) to configure the WebDriverAgent package for a real device.
+
+The `preInstalledWDABundleId` is usually `<bundle id>.xctrunner`. The `<bundle id>` could depend on your provisioning profile environment.
+If it is a non-paid account by `appium` user name, the bundle id would have `com.appium` prefix.
+Then, the WebDriverAgent bumdle id could be `com.appium.WebDriverAgentRunner` for example.
+The test bundle by Xcode will be `com.appium.WebDriverAegnt.xctrunner`.
+The `preInstalledWDABundleId` needs to be the `com.appium.WebDriverAegnt.xctrunner`.
+
+> **Note**
+> Please make sure if the installed WebDriverAgent is valid if the session start continued failing with a correct WebDriverAgent bundle id.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:58]: Please make sure that the installed WebDriverAgent-Runner application is still launchable if the test session startup still fails by providing a correct WebDriverAgent bundle identifier.

File: docs/run-preinstalled-wda.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+> For example, non-paid account has limited period to keep the provisiong profile valid. Sometimes it requires to be uninstalled once, or restart the device.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:59]: Sometimes it is necessary to reinstall WebDriverAgent-Runner once, or to restart the device

File: docs/run-preinstalled-wda.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+#### With 3rd party tools
+
+Some 3rd party tools such as [ios-deploy](https://github.com/ios-control/ios-deploy), [go-ios](https://github.com/danielpaulus/go-ios) and [tidevice](https://github.com/alibaba/taobao-iphone-device) can install the built WebDriverAgent package.
+
+`WebDriverAgentRunner-Runner.app` package may exist in a `derivedDataPath` directory explained in [Real Device Configuration tutorial](./real-device-config.md).
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:65]: as explained in

File: docs/run-preinstalled-wda.md [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+The `WebDriverAgentRunner-Runner.app` can be installed without xcodebuild with the 3rd party tools.
+
+## Note
+
+`appium:webDriverAgentUrl` is prior than `appium:usePreinstalledWDA` since the `appium:webDriverAgentUrl` expects the device under test already has running WebDriverAgent process.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:70]: Could you please explain what this sentence is trying to say?

File: docs/wda-custom-server.md [hunk 18]
```diff
@@ -1,5 +1,5 @@
 ---
-title: WebDriverAgent
+title: Manage WebDriverAgent by yourselves
```
Comment (mykola-mokhnach) [inline docs/wda-custom-server.md:2]: by yourself

File: lib/driver.js (hunk 7 above)
Comment (mykola-mokhnach) [reply lib/driver.js:843]: yes, something like

```
throw new Error(
  `Unable to launch WebDriverAgent because of failure: ${err.message}. ` +
  `Please make sure if the ${this.wda.bundleIdForXctest} exists and it is launchable. ` +
  `${WDA_REAL_DEV_TUTORIAL_URL} may help to complete the preparation.`
)
```

Comment (KazuCocoa) [reply lib/driver.js]: got it

File: lib/driver.js [hunk 19]
```diff
@@ -980,7 +1007,7 @@ class XCUITestDriver extends BaseDriver {
           this.log.debug(`Unable to DELETE session on WDA: '${err.message}'. Continuing shutdown.`);
         }
       }
-      if (!this.wda.webDriverAgentUrl && this.opts.useNewWDA) {
+      if ((!this.wda.webDriverAgentUrl && this.opts.useNewWDA) || this.opts.usePreinstalledWDA) {
```
Comment (KazuCocoa) [inline lib/driver.js:1010]: note:
replace `this.opts.usePreinstalledWDA` with `this.wda.canSkipXcodebuild` if https://github.com/appium/WebDriverAgent/pull/691 is faster than this pr

File: docs/wda-custom-server.md (hunk 18 above)
Comment (KazuCocoa) [inline docs/wda-custom-server.md:2]:
```suggestion
title: Manage WebDriverAgent by yourself
```

File: docs/run-preinstalled-wda.md (hunk 10 above)
Comment (KazuCocoa) [reply docs/run-preinstalled-wda.md]: I removed this sentence to avoid possible confusion.

**Comment by KazuCocoa** [comment]
Updated. tweaked wordings such as Appium -> XCUITest driver as well.

File: docs/capabilities.md [hunk 20]
```diff
@@ -56,6 +56,7 @@ Capability | Description
 |`appium:showXcodeLog`|Whether to display the output of the Xcode command used to run the tests. If this is `true`, there will be **lots** of extra logging at startup. Defaults to `false`|e.g., `true`|
 |`appium:iosInstallPause`|Time in milliseconds to pause between installing the application and starting `WebDriverAgent` on the device. Used particularly for larger applications. Defaults to `0`|e.g., `8000`|
 |`appium:usePrebuiltWDA`|Skips the build phase of running the WDA app. Building is then the responsibility of the user. Only works for Xcode 8+. Defaults to `false`.|e.g., `true`|
+|`appium:usePreinstalledWDA`| Whether to launch a preinstalled WebDriverAgentRunner application instead of running `xcodebuild` to start a WebDriverAgentRunner process for real devices. The preinstalled WebDriverAgent package must have built with over Xcode 12. The default target bundle identifier is `com.facebook.WebDriverAgentRunner.xctrunner`. If `appium:updatedWDABundleId` is provided, `<appium:updatedWDABundleId>.xctrunner` will be launched. Please read [Run Preinstalled WebDriverAgentRunner](./run-preinstalled-wda.md) for more details. Defaults to `false`. |`true` or `true`|
```
Comment (mykola-mokhnach) [inline docs/capabilities.md:59]: must have been built by Xcode 12+

File: docs/run-preinstalled-wda.md [hunk 21]
```diff
@@ -0,0 +1,66 @@
+---
+title: Run Preinstalled WebDriverAgentRunner
+---
+
+XCUITest driver can launch preinstalled WebDriverAgent directly against a real device.
+It lets you start an XCUITest driver session without the `xcodebuild` command execution to improve the session start performance.
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply docs/run-preinstalled-wda.md:6]: start performance > startup performance

File: docs/run-preinstalled-wda.md [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+### Capabilities
+
+- Required
+  - `appium:usePreinstalledWDA` to start a WebDriverAgent Runner-process without xcodebuild
+- Optional
+  - `appium:updatedWDABundleId` to customize the WebDriverAgent bundle id XCUITest driver launches
+
+### Example steps with Xcode
+
+1. Run WebDriverAgent with Xcode as test
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:19]: This probably needs some more detailed description, like:

- Open WebDriverAgent in Xcode
- Select WebDriverAgentRunner project
- Choose the target device
- Select Product->Test from the main menu bar

File: docs/run-preinstalled-wda.md [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+Running test of [WebDriveragent](https://github.com/appium/WebDriverAgent) package with Xcode is the easiest way to prepare the device environment.
+Please read [Real Device Configuration tutorial](real-device-config.md) to configure the WebDriverAgent package for a real device.
+
+If it is a non-paid account by `appium` user name, the bundle id would have `com.appium` prefix.
+Then, the WebDriverAgent-Runner's bumdle id could be `com.appium.WebDriverAgentRunner` for example.
+`appium:updatedWDABundleId` value should be `com.appium.WebDriverAgentRunner` then.
+The test bundle by Xcode will be `com.appium.WebDriverAegnt.xctrunner`.
+
+> **Note**
+> Older than Xcode 11 has different naming convension. This feature does not work for a package which is built by Xcode 11 and lower versions.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:55]: convension -> convention

File: lib/driver.js (hunk 19 above)
Comment (mykola-mokhnach) [reply lib/driver.js:1010]: where do we actually perform capabilities validation (e.g. that the target device is not a simulator)?

Comment (KazuCocoa) [reply lib/driver.js:1010]: Perhaps `const usePreinstalledWDA = this.isRealDevice() && this.opts.usePreinstalledWDA;` in L751 is better. I'll add it

File: docs/run-preinstalled-wda.md (hunk 22 above)
Comment (KazuCocoa) [reply docs/run-preinstalled-wda.md:24]: Added to guide to With Xcode below and it has more concrete steps

File: lib/driver.js (hunk 19 above)
Comment (KazuCocoa) [reply lib/driver.js:1013]: added https://github.com/appium/appium-xcuitest-driver/pull/1609/files#diff-ed3c013edbc896bd87ef62539626fcdc8930a31f3cee7ecb37438e172af2c8e0R751

File: docs/capabilities.md [hunk 24]
```diff
@@ -56,6 +56,7 @@ Capability | Description
 |`appium:showXcodeLog`|Whether to display the output of the Xcode command used to run the tests. If this is `true`, there will be **lots** of extra logging at startup. Defaults to `false`|e.g., `true`|
 |`appium:iosInstallPause`|Time in milliseconds to pause between installing the application and starting `WebDriverAgent` on the device. Used particularly for larger applications. Defaults to `0`|e.g., `8000`|
 |`appium:usePrebuiltWDA`|Skips the build phase of running the WDA app. Building is then the responsibility of the user. Only works for Xcode 8+. Defaults to `false`.|e.g., `true`|
+|`appium:usePreinstalledWDA`| Whether to launch a preinstalled WebDriverAgentRunner application instead of running `xcodebuild` to start a WebDriverAgentRunner process for real devices. The preinstalled WebDriverAgent package must have been built by Xcode 12+. The default target bundle identifier is `com.facebook.WebDriverAgentRunner.xctrunner`. If `appium:updatedWDABundleId` is provided, `<appium:updatedWDABundleId>.xctrunner` will be launched. Please read [Run Preinstalled WebDriverAgentRunner](./run-preinstalled-wda.md) for more details. Defaults to `false`. |`true` or `true`|
```
Comment (mykola-mokhnach) [inline docs/capabilities.md:59]: preinstalled WebDriverAgentRunner application using a custom XCTest API client instead

Comment (mykola-mokhnach) [reply docs/capabilities.md:59]: `to start a WebDriverAgentRunner process` is redundant

Comment (mykola-mokhnach) [reply docs/capabilities.md:59]: `true` or `true` -> typo?

File: docs/run-preinstalled-wda.md [hunk 25]
```diff
@@ -0,0 +1,81 @@
+---
+title: Run Preinstalled WebDriverAgentRunner
+---
+
+XCUITest driver can launch preinstalled WebDriverAgent directly against a real device.
+It lets you start a XCUITest driver session without the `xcodebuild` command execution to improve the session startup performance.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:6]: lets you to start

File: docs/run-preinstalled-wda.md [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+
+## For Real Devices
+
+### Capabilities
+
+- Required
+  - `appium:usePreinstalledWDA` to start a WebDriverAgent Runner-process without xcodebuild
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:13]: Here we can simply links to capabilities document

File: docs/run-preinstalled-wda.md [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+- Optional
+  - `appium:updatedWDABundleId` to customize the WebDriverAgent bundle id XCUITest driver launches
+
+### Example steps with Xcode
+
+Open WebDriverAgent in Xcode
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:19]: Please merge these steps with the below one into a single numbered list.
We could also mention that the `appium driver run xcuitest open-wda` command could be used to quickly open WDA in Xcode

File: docs/run-preinstalled-wda.md [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+> **Note**
+> Please ensure that the WebDriverAgent-Runner application is launchable before starting a XCUITest driver session.
+> For example, whether the provisioning profile is trusted by the system.
+
+### Install WebDriverAgent
+
+#### With Xcode
+
+Running test of [WebDriveragent](https://github.com/appium/WebDriverAgent) package with Xcode is the easiest way to prepare the device environment.
+The step is:
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:53]: The steps are:

File: docs/run-preinstalled-wda.md [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+
+1. Open WebDriverAgent project in Xcode
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:55]: these steps repeat the below ones, aren't they?

File: lib/driver.js [hunk 30]
```diff
@@ -747,12 +747,18 @@ class XCUITestDriver extends BaseDriver {
           `to avoid conflicts and speed up the building process`
       );
     }
+
+    if (this.opts.usePreinstalledWDA && !this.isRealDevice()) {
+      throw new Error(`'usePreinstalledWDA' capability must be against a real device`);
```
Comment (mykola-mokhnach) [inline lib/driver.js:752]: ...capability is only supported for real devices. Use `appium:useXctestrunFile` one  if you need to init WebDriverAgent without xcodebuild on Simulators

File: lib/driver.js [hunk 31]
```diff
@@ -747,12 +747,18 @@ class XCUITestDriver extends BaseDriver {
...
+    }
+    const usePreinstalledWDA = this.isRealDevice() && this.opts.usePreinstalledWDA;
```
Comment (mykola-mokhnach) [inline lib/driver.js:754]: here this.isRealDevice() check is redundant

File: docs/run-preinstalled-wda.md (hunk 27 above)
Comment (KazuCocoa) [reply docs/run-preinstalled-wda.md:19]: oops

File: lib/driver.js (hunk 30 above)
Comment (KazuCocoa) [reply lib/driver.js:752]: `useXctestrunFile` uses xcodebuild command, but it uses an existing built package. So it is not equivalent to usePreinstalledWDA. (so it is available for real devices as well)
Maybe useXctestrunFile is similar to usePrebuiltWDA capability but more portable

Comment (mykola-mokhnach) [reply lib/driver.js]: so there is no direct equivalent for simulators?

Comment (KazuCocoa) [reply lib/driver.js]: correct, I haven't seen such a method yet. i haven't investigated well, but I wondered if we could use simctl something like https://stackoverflow.com/questions/33633882/run-xctest-bundle-with-simctl , but maybe need additional argument etc

**Comment by KazuCocoa** [comment]
Merged from master

File: docs/run-preinstalled-wda.md [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+
+Please read [Real Device Configuration tutorial](real-device-config.md) to configure the WebDriverAgent package for a real device before the step 4.
+
+If it is a non-paid account by `appium` user name, the bundle id would have `com.appium` prefix.
+Then, the WebDriverAgent-Runner's bumdle id could be `com.appium.WebDriverAgentRunner` for example.
+`appium:updatedWDABundleId` value should be `com.appium.WebDriverAgentRunner` then.
+The test bundle by Xcode will be `com.appium.WebDriverAegnt.xctrunner`.
+
+> **Note**
+> Older than Xcode 11 has different naming convention. This feature does not work for a package which is built by Xcode 11 and lower versions.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:71]: Xcode versions below 12 have different naming conventions.

File: docs/run-preinstalled-wda.md [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+
+> **Note**
+> Please make sure that the installed `WebDriverAgentRunner-Runner` application is still launchable if the XCUITest driver session startup continued still fails by providing a correct WebDriverAgent bundle identifier.
```
Comment (mykola-mokhnach) [inline docs/run-preinstalled-wda.md:74]: if the XCUITest driver session startup still fails by providing a correct WebDriverAgent bundle identifier

File: mkdocs.yml [hunk 34]
```diff
@@ -40,6 +40,8 @@ nav:
       - troubleshooting.md
       - reference/commands/appium-xcuitest-driver.md
   - Guides:
+      - Improve Setup performance:
```
Comment (mykola-mokhnach) [inline mkdocs.yml:43]: setup or startup?

**Review by mykola-mokhnach — APPROVED** [review]
